\chapter{System Implementation and Testing}

This chapter describes the implementation and evaluation of the self-documenting DSL designed for transforming SQL tables into RDF\@.

\section{Hardware and Software Requirements}

\subsection{Hardware Requirements}

The development of this DSL was conducted on a server provided by the University of Tennessee Health Science Centre.  This server had the following specifications:

\begin{center}
\begin{tabular}{ll}
CPU & Dual 32-core AMD EPYC 7551 processors\\
Kernel & Linux Kernel version 4.9.0{-}14{-}amd64 x86\_64\\
Memory & 251.5939 GiB of RAM\\
Storage & 24.22 TiB\\
Operating System & Debian GNU/Linux 9.13 (stretch)\\
\end{tabular}
\end{center}

\subsection{Software Requirements}

The software dependencies for this DSL was managed using GNU Guix\footnote{https://guix.gnu.org/}.  The implementation was done in a scheme dialect called GNU Guile\footnote{https://www.gnu.org/software/guile/}.  The database that was used was MariaDB and the RDF datastore utilised was Virtuoso\footnote{https://virtuoso.openlinksw.com/}.

\section{DSL Implementation}

This DSL uses Scheme's flexible hygienic macro system to offer end-users a human-friendly syntax for interacting with SQL databases and generating RDF output bundled with transparent documentation.  The implementation of this DSL involves several key components, as illustrated in Figure~\ref{fig:system-diagram} that include: defining ontologies, defining transformers, processing SQL queries, mapping SQL results to RDF, and generating documentation.  The subsections below outline the implementation of these various parts.

\subsection{Defining Transformers}

A special syntax, \codeword{``define-transformer''} is introduced.  This special syntax transforms a view of a database.\   \codeword{``define-transformer''} consists of three order-agnostic clauses: \codeword{``tables''}, \codeword{``schema-triples''} and \codeword{``triples''}, in the form show in Figure~\ref{code:define-transformer-syntax} below:

\begin{figure}[H]
\centering
\begin{minted}{scheme}
(define-transformer function-name
  (tables (table ...) raw-forms ...)
  (schema-triples
   (subject predicate object) ...)
  (triples subject
    (verb predicate object) ...))
\end{minted}
\caption[\textit{``define-transformer''} special form]{\textbf{``define-transformer'' syntax.}  This shows the general form of the ``define-transformer'' syntax.  It comprises the ``tables'', ``schema-triples'' and ``triples'' form.}\label{code:define-transformer-syntax}
\end{figure}

The \codeword{``tables''} clause specifies the database tables to be joined to construct the view to be transformed.\   \textit{``table''} can be either of the form \textit{``table-name''} or of the form: \codeword{``(JOIN-OPERATOR TABLE-NAME RAW-CONDITION)''}.\   \textit{``table-name''} is the name of the table.\   \codeword{``JOIN-OPERATOR''} can be one of: \codeword{``join''}, \codeword{``left-join''} and \codeword{``inner-join''}.\   \textit{``RAW-CONDITION''} is the join condition as a raw string.  This is usually something like \codeword{``USING (SpeciesId)''}.\   \textit{``raw-forms''} are expressions that must evaluate to strings to be appended to the SQL query.  The example below shows an example of doing a left join between a Genbank table with a Species table using the ``SpeciesId'' as the foreign key as show in Figure~\ref{code:genbank-species-join-example}:

\begin{figure}[H]
\centering
\begin{minted}{scheme}
(tables (Genbank
           (left-join Species "USING (SpeciesId)")))
\end{minted}
\caption[\textit{``table''} special form]{\textbf{Using the ``table'' special form to select data from tables}.  This example use the ``table'' special form to do a left join between the \textit{Genbank} and \textit{Species} table using the \textit{SpeciesId} column as the foreign key.}\label{code:genbank-species-join-example}
\end{figure}

The \codeword{``schema-triples''} clause specifies the list of triples to be written once when the transform starts.  An example of how this would like is demonstrated in Figure~\ref{code:schema-triples-example}:

\begin{figure}[H]
\centering
\begin{minted}{scheme}
(schema-triples
   (gnc:nucleotide a skos:Concept)
   (gnt:hasSequence rdfs:domain gnc:nucleotide))
\end{minted}
\caption[\textit{``schema-triples''} special form]{\textbf{Defining special triples that will be run only once before the SQL query is executed.}  This example defines 2 triples.  These 2 triples are are generated only once for the entire SQL query.}\label{code:schema-triples-example}
\end{figure}

On the other hand, the \codeword{``triples''} clause specifies the triples to be transformed once for each row in the view.  All triples have a common \textit{``subject''}.  The \codeword{``(verb predicate object)''} clauses are described below:

\textit{``verb''} can either be a \codeword{``set''} or \codeword{``multiset''}.  For the \codeword{``set''} \textit{``verb''}, a single triple \codeword{``(SUBJECT PREDICATE OBJECT-VALUE)''} is written where \textit{``OBJECT-VALUE''} is the result of evaluating \textit{``OBJECT''}.  For the \codeword{``multiset''} \textit{``verb''}, \textit{``OBJECT''} evaluates to a list, and a triple \codeword{``(SUBJECT PREDICATE OBJECT-VALUE-ELEMENT)''} is created for each element \textit{``OBJECT-VALUE-ELEMENT''} of that list.

The \textit{``subject''} and \textit{``object''} expressions in the triples clause reference database fields using a \codeword{``(field TABLE COLUMN)''} clause where \textit{``TABLE''} and \textit{``COLUMN''} refer to the table and column of the field being referenced.  Database fields can also be referenced using \codeword{``(field TABLE COLUMN ALIAS)''} where \textit{``ALIAS''} is an alias for that column in the SQL query.

The example in Figure~\ref{code:triples-example} below show's how a triple can be defined:

\begin{figure}[H]
\centering
\begin{minted}{scheme}
(triples (field Genbank Id)
    (set gnt:hasSequence (field Genbank Sequence))
    (set gnt:belongsToSpecies (field Species Fullname)))
\end{minted}
\caption[\textit{``triples''} special form]{\textbf{Defining triples from table rows.}  This example shows how triples are expressed declaratively from the relevant table fields.  Each row returned from the SQL query will be parsed into the triples matching the form expressed here.}\label{code:triples-example}
\end{figure}

\subsection{Executing a Transformer}

To execute transformers, an extra special syntax, \codeword{``with-documentation''} is provided.  This special syntax is shown in Figure~\ref{code:with-documentation-syntax} below:

\begin{figure}[H]
\centering
\begin{minted}{scheme}
(with-documentation
   (name ...)
   (connection ...)
   (table-metadata? ...)
   (prefixes ...)
   (inputs ...)
   (outputs
    `(#:documentation ...
      #:rdf ...)))
\end{minted}
\caption[\textit{``with-documentation''} special form]{\textbf{The general form of the ``with-documentation'' syntax.}  This special form glues together: defined transformers, RDF prefixes, important connection parameters to the Virtuoso store and specifies where the output is generated to.}\label{code:with-documentation-syntax}
\end{figure}

Each part of the \codeword{``with-documentation''} is described below:

\begin{enumerate}
\item \codeword{``name''}: This represents the title used in the generated documentation markdown file.
\item \codeword{``connection''}: This contains the SQL configuration variables that will be used to connect to the database.
\item \codeword{``prefixes''}: These are various ontologies that will be included at the beginning of the RDF document.
\item \codeword{``inputs''}: This is a list of defined transformers that will be executed.
\item \codeword{``outputs''}: This specifies the file path where the RDF and documentation files will be saved.
\end{enumerate}

A more complete example is shown in Figure~\ref{code:with-documentation-example} below:

\begin{figure}[H]
\centering
\begin{minted}{scheme}
(with-documentation
   (name "Genebank Metadata")
   (connection %connection-settings)
   (table-metadata? #f)
   (prefixes
    '(("rdf:" "<http://www.w3.org/1999/02/22-rdf-syntax-ns#>")
      ("rdfs:" "<http://www.w3.org/2000/01/rdf-schema#>")))
   (inputs
    (list
     genbank))
   (outputs
    `(#:documentation "/tmp/genbank.md"
      #:rdf "/tmp/genbank.ttl")))
\end{minted}
\caption[Transforming Genbank SQL data into RDF using \textit{``with-documentation''} special form]{\textbf{Transforming Genbank SQL data into RDF}.  This example uses the ``with-documentation'' special form to convert data from the Genbank table to RDF.  We use the prefixes ``rdf:'' and ``rdfs:''.  As input, we use the genbank transformer.  The auto-generated document is stored in ``/tmp/genbank.md'' and the generated turtle file is stored in ``/tmp/genbank.ttl''.  In this example we don't include table metadata provided from the SQL engine in ``/tmp/genbank.ttl'' by setting the ``table-metadata'' option to ``false''.  Connection settings to the MariaDB is provided by ``\%connection-settings''.}\label{code:with-documentation-example}
\end{figure}

\section{System Testing and Validation}

In this section, various tests were done to assess the performance and functionality of the self-documenting DSL designed for transformed tables into RDF format.  In addition to assessing the performance of the SPARQL queries from the generated RDF, the quality of the output is also assessed, and federated queries that are impossible to do in SQL are highlighted.

\subsection{Query Quality and Execution Performance}
\label{sec:benchmark}

The following SQL query, shown in Figure~\ref{sql:counting-pubs-with-diabetes} below, was run several times to count the number of publications in the Genenetwork SQL store related to diabetes:

\begin{figure}[H]
\centering
\begin{minted}{sql}
  SELECT COUNT(DISTINCT Id) FROM Publication
  WHERE Abstract LIKE "%diabetes%";
\end{minted}
\caption[SQL Query to count publications with the word ``diabetes'']{\textbf{Finding the number of Publications whose Abstract have the word ``diabetes'' from the GN database.}  This query shows the number of distinct publications tha contain the word ``diabetes'' in their abstract from the GN SQL database.}\label{sql:counting-pubs-with-diabetes}
\end{figure}

Similarly, Figure~\ref{sparql:counting-pubs-with-diabetes} below shows the corresponding SPARQL query use to count the number of publications in RDF\@.

\begin{figure}[H]
\centering
\begin{minted}{sparql}
PREFIX   dct: <http://purl.org/dc/terms/>
PREFIX   rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX fabio: <http://purl.org/spar/fabio/>

SELECT COUNT(DISTINCT ?paper) WHERE {
       ?paper rdf:type fabio:ResearchPaper ;
              dct:abstract ?abstract .
       ?abstract bif:contains "diabetes" .
}
\end{minted}
\caption[SPARQL Query to count publications with the word ``diabetes'']{\textbf{Finding the number of publications whose abstract have the word ``diabetes'' from Virtuoso in SPARQL.}  This query shows the number of distinct publications tha contain the word ``diabetes'' in their abstract using SPARQL.}\label{sparql:counting-pubs-with-diabetes}
\end{figure}

In the 2 queries above, we get 22 publications that have the word ``diabetes'' in their abstract.  The performance metrics of these 2 queries is shown in Table~\ref{table:perf-1} below:

\begin{table}[H]
\begin{tabular}{r|rr|r}
& SQL (seconds) & SPARQL (seconds) & \% perf. difference\\[0pt]
\toprule
& 0.09 & 0.014 & 84.44\\[0pt]
& 0.08 & 0.017 & 78.75\\[0pt]
& 0.10 & 0.016 & 84.00\\[0pt]
& 0.11 & 0.014 & 87.27\\[0pt]
& 0.09 & 0.015 & 83.33\\[0pt]
\toprule
\textit{Mean:} & 0.094 & 0.0152 & 83.33\\[0pt]
\end{tabular}
\caption[Performance comparison in seconds: publication count with "diabetes" (SQL vs SPARQL)]{\textbf{Comparing how fast it takes to get the number of publications with the word ``diabetes'' in the abstract.}}\label{table:perf-1}
\end{table}

The following query, shown in Figure~\ref{sql:counting-phenotypes-with-diabetes}, is more complicated and counts the number of phenotypes that are related to diabetes:

\begin{figure}[H]
\centering
\begin{minted}{sql}
SELECT
  COUNT(DISTINCT Phenotype.Id)
FROM
  PublishXRef
  LEFT JOIN Publication ON
    Publication.Id = PublishXRef.PublicationId 
  LEFT JOIN Phenotype ON
    Phenotype.Id = PublishXRef.PhenotypeId
WHERE
  Publication.Abstract LIKE "%diabetes%";
\end{minted}
\caption[SQL Query to count the number of phenotypes with the word ``diabetes'']{\textbf{Finding the number of phenotypes related to diabetes from the GN database.}  The SQL query is more complex because it requires multiple table joins (I.e., PublishXRef, Publication, and Phenotype) and prior knowledge of the schema to understand the relationships.  Without context, it's difficult to know what each table represents.}\label{sql:counting-phenotypes-with-diabetes}
\end{figure}

The corresponding SPARQL query, shown in Figure~\ref{sparql:counting-phenotypes-with-diabetes}, is:

\begin{figure}[H]
\centering
\begin{minted}{sparql}
PREFIX   dct: <http://purl.org/dc/terms/>
PREFIX   rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX   gnc: <http://genenetwork.org/category/>
PREFIX fabio: <http://purl.org/spar/fabio/>

SELECT COUNT(DISTINCT ?phenotype) WHERE {
       ?paper rdf:type fabio:ResearchPaper ;
              dct:abstract ?abstract .
       ?abstract bif:contains "diabetes" .
       ?phenotype rdf:type gnc:Phenotype ;
                  dct:isReferencedBy ?paper .
}
\end{minted}
\caption[SPARQL Query to count the number of phenotypes with the word ``diabetes'']{\textbf{Finding the number of phenotypes related to diabetes from the GN database.}  This SPARQL query is simpler and less complex compared to it's alternative SQL query---Figure~\ref{sql:counting-phenotypes-with-diabetes}---because it directly expresses relationships using intuitive RDF triples and ontology terms like \textit{fabio:ResearchPaper} and \textit{gnc:Phenotype}.  This makes it easier to understand without needing prior knowledge of the database schema or explicit table joins, unlike the SQL query, which requires joining multiple tables and navigating their relationships manually.}\label{sparql:counting-phenotypes-with-diabetes}
\end{figure}

In both queries above, we get 105 phenotypes that are related to diabetes.  The performance of these 2 queries is tabulated in Table~\ref{table:perf-2} below:

\begin{table}[H]
\begin{tabular}{r|rr|r}
& SQL (seconds) & SPARQL (seconds) & \% perf. difference\\[0pt]
\toprule
& 0.19 & 0.013 & 93.16\\[0pt]
& 0.18 & 0.015 & 91.67\\[0pt]
& 0.17 & 0.014 & 91.76\\[0pt]
& 0.18 & 0.016 & 91.11\\[0pt]
& 0.17 & 0.009 & 94.71\\[0pt]
\toprule
\textit{mean} & 0.178 & 0.0134 & 92.47\\[0pt]
\end{tabular}
\caption[Performance comparison in seconds: phenotype datasets related to diabetes (SQL vs SPARQL)]{\textbf{Comparing how fast it takes to get the number of phenotype datasets with the word ``diabetes''.}}\label{table:perf-2}
\end{table}

A more complicated query that counts the number of datasets across different InbredSet group that have phenotypes related to diabetes is shown in Figure~\ref{sql:datasets-phenotypes-diabetes}.

\begin{figure}[H]
\centering
\begin{minted}{sql}
SELECT 
  COUNT(DISTINCT InfoFiles.InfoFileId) 
FROM 
  InfoFiles 
  LEFT JOIN PublishFreeze
    ON InfoFiles.InfoPageName = PublishFreeze.Name 
  LEFT JOIN Datasets USING (DatasetId)
  LEFT JOIN InbredSet
    ON InfoFiles.InbredSetId = InbredSet.InbredSetId 
  LEFT JOIN PublishXRef
    ON PublishXRef.InbredSetId = InfoFiles.InbredSetId 
  LEFT JOIN Publication
    ON Publication.Id = PublishXRef.PublicationId 
  LEFT JOIN Phenotype ON Phenotype.Id = PublishXRef.PhenotypeId 
WHERE 
  Publication.Abstract LIKE "%diabetes%";
\end{minted}
\caption[SQL query to count diabetes-related phenotype datasets across different inbredset groups]{\textbf{SQL Query to count the number of datasets across different inbredset groups that have phenotypes related to diabetes.}}\label{sql:datasets-phenotypes-diabetes}
\end{figure}

The corresponding SPARQL query is shown in Figure~\ref{sparql:datasets-phenotypes-diabetes}.

\begin{figure}[H]
\centering
\begin{minted}{sparql}
PREFIX   dct: <http://purl.org/dc/terms/>
PREFIX   rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX fabio: <http://purl.org/spar/fabio/>
PREFIX   gnc: <http://genenetwork.org/category/>
PREFIX   gnt: <http://genenetwork.org/term/>
PREFIX  dcat: <http://www.w3.org/ns/dcat#>
PREFIX  xkos: <http://rdf-vocabulary.ddialliance.org/xkos#>

SELECT COUNT(DISTINCT ?dataset) WHERE {
       ?paper rdf:type fabio:ResearchPaper ;
              dct:abstract ?abstract .
       ?abstract bif:contains "diabetes" .
       ?phenotype rdf:type gnc:Phenotype ;
                  gnt:belongsToGroup ?group ;
                  dct:isReferencedBy ?paper .
       ?dataset rdf:type dcat:Dataset ;
                gnt:belongsToGroup ?group .
                  
}
\end{minted}
\caption[SPARQL query to count diabetes-related phenotype datasets across different inbredset group]{\textbf{SPARQL Query to count the number of datasets across different inbredset groups that have phenotypes related to diabetes.}}\label{sparql:datasets-phenotypes-diabetes}
\end{figure}

From the above 2 queries - Figure~\ref{sql:datasets-phenotypes-diabetes} and Figure~\ref{sparql:datasets-phenotypes-diabetes} - we get 2 different dataset counts.  The SQL query returns 406 datasets, while the SPARQL query returns 419 datasets.  The performance metrics of thes 2 queries are tabulated in Table~\ref{table:perf-3} below.

\begin{table}[H]
\begin{tabular}{r|rr|r}
& SQL (seconds) & SPARQL (seconds) & \% perf. difference\\[0pt]
\toprule
& 38.33 & 0.649 & 98.31 \\[0pt]
& 38.46 & 0.124 & 99.68\\[0pt]
& 41.53 & 0.430 & 98.96\\[0pt]
& 38.53 & 0.118 & 99.69\\[0pt]
& 38.50 & 0.696 & 98.19\\[0pt]
\toprule
\textit{mean:} & 39.07 & 0.4034 & 98.97\\[0pt]
\end{tabular}
\caption[Performance comparison in seconds: Counting diabetes-related phenotype datasets across different inbredset group]{\textbf{Comparing how fast it takes to get the number of datasets across different inbredset groups that have phenotypes related to diabetes.}}\label{table:perf-3}
\end{table}

\subsection{Query Output Quality}

Consider the SQL query, shown in Figure~\ref{sql:mouse}, which fetches all the details about the Mouse species from the SQL store.

\begin{figure}[H]
\centering
\begin{minted}{SQL}
SELECT * FROM Species WHERE Name="Mouse" \G
\end{minted}
\caption[SQL: Fetching mouse details from Species table]{\textbf{SQL query to retrieve mouse details from Species table.}}\label{sql:mouse}
\end{figure}

The corresponding results are illustrated in Figure~\ref{sql:mouse-results}.

\begin{figure}[H]
\centering
\begin{minted}{rst}
           Id: 1
    SpeciesId: 1
  SpeciesName: Mouse
         Name: mouse
     MenuName: Mouse (Mus musculus, mm10)
     FullName: Mus musculus
       Family: Vertebrates
FamilyOrderId: 1
   TaxonomyId: 10090
      OrderId: 15
\end{minted}
\caption[SQL Results: Fetching mouse details from Species table]{\textbf{Mouse details from Species table.}  This example illustrates the SQL tabular result structure, which lacks semantic context for the columns unless defined elsewhere.}\label{sql:mouse-results}
\end{figure}

The corresponding sparql query is shown in Figure~\ref{sparql:mouse}:

\begin{figure}[H]
\centering
\begin{minted}{sparql}
PREFIX gnt: <http://genenetwork.org/term/>

CONSTRUCT {
  ?species ?predicate ?object
} WHERE {
  ?species gnt:shortName "mouse";
           ?predicate ?object .
}
\end{minted}
\caption[SPARQL: Fetching mouse details]{SPARQL SELECT query to retrieve mouse details from Species table.}\label{sparql:mouse}
\end{figure}

The enriched results of Figure~\ref{sparql:mouse} are illustrated in Figure~\ref{figure:mouse-sparql-results}

\begin{figure}[H]
\centering
\begin{minted}[breaklines]{json-ld}
{
  "@context": {
    "gnc": "http://genenetwork.org/category/",
    "gnt": "http://genenetwork.org/term/",
    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#>",
    "label": "rdfs:label",
    "prefLabel": "http://www.w3.org/2004/02/skos/core#prefLabel",
    "taxonomy-id": "http://www.w3.org/2004/02/skos/core#notation",
    "family": "gnt:family",
    "shortname": "gnt:shortName",
    "definedBy": "http://www.w3.org/1999/02/22-rdf-syntax-ns#isDefinedBy",
    "in-scheme": "http://www.w3.org/2004/02/skos/core#inScheme",
    "alt-label": "http://www.w3.org/2004/02/skos/core#altLabel"
  },
  "@id": "http://genenetwork.org/id/Mus_musculus",
  "family": "Vertebrates",
  "shortname": "mouse",
  "definedBy": {
    "@id": "http://www.wikidata.org/entity/Q83310"
  },
  "label": "Mus musculus",
  "alt-label": "Mouse",
  "in-scheme": {
    "@id": "gnc:ResourceClassificationScheme"
  },
  "taxonomy-id": {
    "@id": "http://purl.uniprot.org/taxonomy/10090"
  },
  "prefLabel": "Mouse (Mus musculus, mm10)"
}
\end{minted}
\caption[SPARQL Results: Fetching mouse details]{\textbf{Mouse details in json-ld format from Genenetwork Graph in Virtuoso.}  Here, we can see that json-ld provides richer semantic context by clearly defining relationships and attributes through standardized vocabularies and links to relevant resources.}\label{figure:mouse-sparql-results}
\end{figure}

A common operation done in GeneNetwork is to search GeneRIF entries from NCBI to search given probesets for key terms belonging to a given species.   Figure~\ref{sql:search-ncbi} shows how such a query in SQL looks like:

\begin{figure}[H]
\centering
\begin{minted}[breaklines]{SQL}
SELECT 
  DISTINCT Species.*, GeneRIF_BASIC.* 
FROM 
  Species
  INNER JOIN InbredSet ON InbredSet.SpeciesId = Species.Id 
  INNER JOIN ProbeFreeze ON ProbeFreeze.InbredSetId = InbredSet.Id 
  INNER JOIN Tissue ON ProbeFreeze.TissueId = Tissue.Id 
  INNER JOIN ProbeSetFreeze ON ProbeSetFreeze.ProbeFreezeId = ProbeFreeze.Id 
  INNER JOIN ProbeSetXRef ON ProbeSetXRef.ProbeSetFreezeId = ProbeSetFreeze.Id 
  INNER JOIN ProbeSet ON ProbeSet.Id = ProbeSetXRef.ProbeSetId 
  LEFT JOIN Geno ON ProbeSetXRef.Locus = Geno.Name 
  AND Geno.SpeciesId = Species.Id
  INNER JOIN GeneRIF_BASIC ON GeneRIF_BASIC.symbol = ProbeSet.Symbol 
WHERE 
  GeneRIF_BASIC.comment LIKE "%diabetes%" 
  AND Species.Name = "mouse"
  AND GeneRIF_BASIC.`symbol` = BINARY "Gcg"
ORDER BY 
  GeneRIF_BASIC.comment ASC
LIMIT 
  1\G
\end{minted}
\caption[SQL: Search for GeneRIF entries from NCBI for the keyword ``diabetes'' that belongs to the mouse species]{\textbf{SQL Search for GeneRIF entries from NCBI for the keywords ``diabetes'' that belong to the mouse species}}\label{sql:search-ncbi}
\end{figure}

The results of Figure~\ref{sql:search-ncbi} are shown in Figure~\ref{sql:ncbi-results}.

\begin{figure}[H]
\centering
\begin{minted}{rst}
           Id: 1
    SpeciesId: 1
  SpeciesName: Mouse
         Name: mouse
     MenuName: Mouse (Mus musculus, mm10)
     FullName: Mus musculus
        Family: VertebratesFamilyOrderId: 1
   TaxonomyId: 10090
      OrderId: 15
    SpeciesId: 2
        TaxID: 10116
       GeneId: 24952
       symbol: Gcg
    PubMed_ID: 26037249
   createtime: 2015-10-24 11:29:00
      comment: Data (including data from transgenic rats) suggest 1)
               celiac ganglia neurotransmission and 2) glucagon
               secretion are down-regulated in short-term
               hyperglycemia of diabetes; Wallerian degeneration
               protein mutation protects against these dysfunctions.
    VersionId: 1
1 row in set (0.01 sec)
\end{minted}
\caption[SQL Results: Search for GeneRIF entries from NCBI for the keyword ``diabetes'' that belongs to the mouse species]{Results of SQL search for GeneRIF from NCBI for the keywords ``diabetes'' that belong to the mouse species}\label{sql:ncbi-results}
\end{figure}

The corresponding SPARQL query for Figure~\ref{sql:search-ncbi} is shown in Figure:~\ref{sparql:ncbi-rif}:

\begin{figure}[H]
\centering
\begin{minted}{sparql}
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX gnt: <http://genenetwork.org/term/>
PREFIX gnc: <http://genenetwork.org/category/>
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX gn: <http://genenetwork.org/id/>

CONSTRUCT {
  ?species ?speciesPred ?speciesObj .
  gnt:comment ?commentPred ?commentObj .
} WHERE {
?probeset rdf:type gnc:Probeset ;
          gnt:geneSymbol "Gcg" ;
          ?p ?o .
?symbol rdfs:comment _:node ;
        rdfs:label "Gcg" .
_:node rdf:type gnc:NCBIWikiEntry ;
       gnt:belongsToSpecies ?species ;
       rdfs:comment ?comment ;
       ?commentPred ?commentObj .
?species gnt:shortName "mouse";
       ?speciesPred ?speciesObj .
?comment bif:contains "diabetes" .
}
\end{minted}
\caption[SPARQL: Search for GeneRIF entries from NCBI for the keyword ``diabetes'' that belongs to the mouse species]{\textbf{SPARQL Search for GeneRIF entries from NCBI for the keywords ``diabetes'' that belong to the mouse species}}\label{sparql:ncbi-rif}
\end{figure}

The corresponding results for Figure~\ref{sparql:ncbi-rif} are shown below:

\begin{figure}[H]
\centering
\begin{minted}[breaklines]{json-ld}
{
  "@context": {
    "hasGeneId": {
      "@id": "http://genenetwork.org/term/hasGeneId",
      "@type": "@id"
    },
    "created": {
      "@id": "http://purl.org/dc/terms/created",
      "@type": "http://www.w3.org/2001/XMLSchema#datetime"
    },
    "altLabel": {
      "@id": "http://www.w3.org/2004/02/skos/core#altLabel"
    },
    "inScheme": {
      "@id": "http://www.w3.org/2004/02/skos/core#inScheme",
      "@type": "@id"
    },
    "family": {
      "@id": "http://genenetwork.org/term/family"
    },
    "shortName": {
      "@id": "http://genenetwork.org/term/shortName"
    },
\end{minted}
\end{figure}

\begin{figure}[H]
\centering
\begin{minted}[breaklines, breakanywhere=True]{json-ld}
    "isDefinedBy": {
      "@id": "http://www.w3.org/1999/02/22-rdf-syntax-ns#isDefinedBy",
      "@type": "@id"
    },
    "belongsToSpecies": {
      "@id": "http://genenetwork.org/term/belongsToSpecies",
      "@type": "@id"
    },
    "prefLabel": {
      "@id": "http://www.w3.org/2004/02/skos/core#prefLabel"
    },
    "label": {
      "@id": "http://www.w3.org/2000/01/rdf-schema#label"
    },
    "comment": {
      "@id": "http://www.w3.org/2000/01/rdf-schema#comment"
    },
    "notation": {
      "@id": "http://www.w3.org/2004/02/skos/core#notation",
      "@type": "@id"
    },
    "hasVersionId": {
      "@id": "http://genenetwork.org/term/hasVersionId"
    }
  },
  "@graph": [
    {
      "@id": "http://genenetwork.org/id/Mus_musculus",
      "family": "Vertebrates",
      "shortName": "mouse",
      "isDefinedBy": "http://www.wikidata.org/entity/Q83310",
      "label": "Mus musculus",
      "altLabel": "Mouse",
      "inScheme": "http://genenetwork.org/category/ResourceClassificationScheme",
      "notation": "http://purl.uniprot.org/taxonomy/10090",
      "prefLabel": "Mouse (Mus musculus, mm10)"
    },
    {
      "@id": "http://genenetwork.org/term/comment",
      "@type": "http://genenetwork.org/category/NCBIWikiEntry",
      "belongsToSpecies": "http://genenetwork.org/id/Mus_musculus",
      "hasGeneId": "http://www.ncbi.nlm.nih.gov/gene?cmd=Retrieve&dopt=Graphics&list_uids=14526",
      "hasVersionId": 1,
      "created": "2010-10-23T06:10:00",
\end{minted}
\end{figure}

\begin{figure}[H]
\centering
\begin{minted}[breaklines, breakanywhere=True]{json-ld}
      "comment": {
        "@type": "http://www.w3.org/2001/XMLSchema#string",
        "@value": "Data show that in an animal model of diabetes induced by alloxan, progression of hyperglycemia was significantly attenuated in mice given the Ad-GLP-1 vector compared with control mice."
      },
      "notation": "https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=10090"
    }
  ]
}
\end{minted}
\caption[SPARQL Results: Search for GeneRIF entries from NCBI for the keyword ``diabetes'' that belongs to the mouse species]{\textbf{SPARQL Search for GeneRIF entries from NCBI for the keywords ``diabetes'' that belong to the mouse species}}\label{sparql:ncbi-rif-results}
\end{figure}

From the above results, we see that the SPARQL query results offer richer contextual information.  Each subject and predicate field, analogous to SQL columns, is linked to a resource providing additional definitions.  Furthermore, the versatility of SPARQL extends to the ability to output results to HTML, CSV, TSV, JSON, JSON-LD and other formats~\footnote{http://docs.openlinksw.com/virtuoso/rdfsparql.html\#rdfsparqlimplementatioptragmassfs}.

\subsection{Federated Queries in SPARQL}

SQL is limited to querying tables that reside within the local server environment and does not have the capability to access or query external servers on the internet.  However, RDF overcomes this limitation by allowing the use of federated queries that allow a person to query external services.  The following example, shown in Figure~\ref{sparql:federated}, highlights SQL's limitation.  It queries GeneNetwork to get the Wikidata identifier that can be used to query Wikidata as an external service, to get the longest recorded life span of a mouse which is 4 years.

\begin{figure}[H]
\centering
\begin{minted}{sparql}
PREFIX       gnt: <http://genenetwork.org/term/>
PREFIX       rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX       wdt: <http://www.wikidata.org/prop/direct/>

SELECT ?years {
  ?species gnt:shortName "mouse";
           rdf:isDefinedBy ?id .
SERVICE <https://query.wikidata.org/sparql> {
    ?id wdt:P4214 ?years .
  }
}
\end{minted}
\caption[Federated Query to get longest recorded life span of a mouse from Wikidata]{\textbf{Using a Federated Query.}  In this example, a federated query is used to annotate Genenetwork data using metadata from an external service, Wikidata, to fetch the longest recorded life span of a mouse.}\label{sparql:federated}
\end{figure}

%%% Local Variables:
%%% ispell-local-dictionary: "en_GB-ise"
%%% End:
